storage/src/mmr/iterator.rs

```diff
@@ -109,6 +109,63 @@ pub(crate) fn nodes_needing_parents(peak_iterator: PeakIterator) -> Vec<u64> {
     peaks
 }
 
+/// Returns the position of the oldest provable node in the represented MMR.
+pub(crate) fn oldest_provable_pos(
+    peak_iterator: PeakIterator,
+    oldest_remembered_node_pos: u64,
+) -> u64 {
+    for (peak_pos, height) in peak_iterator {
+        if peak_pos >= oldest_remembered_node_pos {
+            // We have found the tree containing the oldest remembered node. Now we look for the
+            // highest node in this tree whose left-sibling is pruned (if any). The provable nodes
+            // are those that strictly follow this node. If no such node exists, then all existing
+            // nodes are provable
+            let mut two_h = 1 << height;
+            let mut cur_node = peak_pos;
+            while two_h > 1 {
+                let left_pos = cur_node - two_h;
+                let right_pos = left_pos + two_h - 1;
+                if left_pos < oldest_remembered_node_pos {
+                    // found pruned left sibling
+                    return right_pos + 1;
+                }
+                two_h >>= 1;
+                cur_node = left_pos;
+            }
+            return oldest_remembered_node_pos;
+        }
+    }
+    0 // mmr is empty
```

Comment:
nit: I think we should be more careful about this case? Couldn't you also hit here if each `peak_pos` is < `oldest_remembered_node_pos`? I think we should handle those differently.

Reply:
In a valid MMR, the very last node is *always* a peak. So this wouldn't happen, unless the MMR is empty, or oldest_remembered_node_pos is specified incorrectly.

Reply:
I think I'm referring to the latter case. It "shouldn't" happen but I think we should consider returning an error if it does.

Reply:
I see.  I have made it panic(), is that OK vs returning error? I think of this like an array out of bounds condition.

Reply:
if it should never ever happen, I think assert/panic makes sense.